google/cloud/status.cc

```diff
@@ -91,6 +71,119 @@ std::ostream& operator<<(std::ostream& os, StatusCode code) {
   return os << StatusCodeToString(code);
 }
 
+// Encapsulates the implementation of a non-OK status. OK Statuses are
+// represented by a nullptr Status::impl_, as an optimization for the common
+// case of OK Statuses. This class holds all the data associated with a non-OK
+// Status so we don't have to worry about bloating the common OK case.
+class Status::Impl {
+  using PayloadType = std::unordered_map<std::string, std::string>;
+
+ public:
+  explicit Impl(StatusCode code, std::string message, PayloadType payload)
+      : code_(code),
+        message_(std::move(message)),
+        payload_(std::move(payload)) {}
+
+  StatusCode code() const { return code_; }
+  std::string const& message() const { return message_; }
+  PayloadType const& payload() const { return payload_; };
+
+  // Allows mutable access to payload, which is needed in the
+  // `internal::SetPayload()` function.
+  PayloadType& payload() { return payload_; };
+
+  friend inline bool operator==(Impl const& a, Impl const& b) {
+    return a.code_ == b.code_ && a.message_ == b.message_ &&
+           a.payload_ == b.payload_;
+  }
+
+  friend inline bool operator!=(Impl const& a, Impl const& b) {
+    return !(a == b);
+  }
+
+ private:
+  StatusCode code_;
+  std::string message_;
+  PayloadType payload_;
+};
+
+Status::~Status() { delete impl_; }
+
+// Deep copy
+Status::Status(Status const& other)
+    : impl_(other.ok() ? nullptr : new auto(*other.impl_)) {}
+
+// Deep copy
+Status& Status::operator=(Status const& other) {
+  if (this != &other) {
+    delete impl_;
+    impl_ = other.ok() ? nullptr : new auto(*other.impl_);
```

Comment:
this is not exception-safe.  Should `auto(*other.impl_)` throw (or `new auto(...)` throw) the current object was already modified.  That is why we do the copy + swap stuff.

Reply:
Better with unique_ptr?

Reply:
Yes, thanks!